File: packages/python/src/realtime_mail/trust.py
from __future__ import annotations

from .models import MailSource, RealtimeMailMessage, TrustCapability


class TrustPolicy:
    def __init__(self) -> None:
        self._trusted_domains: set[str] = set()

    def trust_domain(self, domain: str) -> None:
        self._trusted_domains.add(domain)

    def revoke_domain(self, domain: str) -> None:
        self._trusted_domains.discard(domain)

    def is_trusted(self, domain: str) -> bool:
        return domain in self._trusted_domains

    def can_render(self, message: RealtimeMailMessage) -> bool:
        if message.source == MailSource.TRADITIONAL:
            return TrustCapability.RENDER_HTML in message.capabilities
        return self.is_trusted(message.domain) and TrustCapability.RENDER_HTML in message.capabilities

    def can_run_script(self, message: RealtimeMailMessage) -> bool:
        return (
            message.source == MailSource.REALTIME
            and self.is_trusted(message.domain)
            and TrustCapability.RUN_SCRIPT_SANDBOXED in message.capabilities
        )

File: packages/go/realtimemail/trust.go
package realtimemail

type TrustPolicy struct {
	trustedDomains map[string]bool
}

func NewTrustPolicy() *TrustPolicy {
	return &TrustPolicy{trustedDomains: map[string]bool{}}
}

func (policy *TrustPolicy) TrustDomain(domain string) {
	policy.trustedDomains[domain] = true
}

func (policy *TrustPolicy) RevokeDomain(domain string) {
	delete(policy.trustedDomains, domain)
}

func (policy *TrustPolicy) IsTrusted(domain string) bool {
	return policy.trustedDomains[domain]
}

func (policy *TrustPolicy) CanRender(message RealtimeMailMessage) bool {
	if message.Source == Traditional {
		return hasCapability(message.Capabilities, RenderHTML)
	}
	return policy.IsTrusted(message.Domain) && hasCapability(message.Capabilities, RenderHTML)
}

func (policy *TrustPolicy) CanRunScript(message RealtimeMailMessage) bool {
	return message.Source == Realtime && policy.IsTrusted(message.Domain) && hasCapability(message.Capabilities, RunScriptSandboxed)
}
